File: android/data/src/main/kotlin/com/ridelink/data/library/LibraryIndexer.kt
package com.ridelink.data.library

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.OpenableColumns
import com.ridelink.core.library.DecodeStatus
import com.ridelink.core.library.IndexReconciliation
import com.ridelink.core.library.LibraryEntry
import com.ridelink.core.library.LocalTrackLocation
import com.ridelink.core.library.MetadataNormalizer
import com.ridelink.core.model.LocalEntryId
import com.ridelink.core.model.QuickId
import com.ridelink.core.model.Track
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import java.security.MessageDigest
import java.util.UUID

/**
 * The whole indexing pipeline (this phase's brief §19): discover → validate supported type →
 * gather basic file info → [com.ridelink.core.model.QuickId] → metadata → artwork → upsert. Every
 * step that can fail does so into a [DecodeStatus], never an exception the caller has to catch —
 * "even local files are untrusted input" (brief §20) applies to every file this touches.
 *
 * **Identity (ADR-005 Amendment A1, this phase's closure-audit CRITICAL finding):** every row's real
 * identity is a freshly-generated [LocalEntryId] — [newLocalEntryId] by default, overridable only for
 * deterministic tests. Reconciliation and every repository write below are keyed by
 * [LocalTrackLocation]/[LocalEntryId], never by [QuickId]: `QuickId` samples only size plus the
 * first/last 64 KiB, so two genuinely different files over 128 KiB can share one, and using it as
 * cross-row identity used to let that silently collapse two different files into one row.
 *
 * Runs on whatever dispatcher the caller supplies (the `app`-layer composition root chooses
 * `Dispatchers.IO`, matching [com.ridelink.network.control.ControlSessionManager]'s convention of
 * deciding dispatchers at the composition root rather than inside a reusable class); every loop
 * checks [ensureActive] so a cancelled scan actually stops rather than merely being ignored (brief
 * §19's "cancellable").
 */
class LibraryIndexer(
    private val context: Context,
    private val repository: LibraryRepository,
    private val artworkCache: ArtworkCache,
    private val monotonicNowUs: () -> Long,
    private val newLocalEntryId: () -> LocalEntryId = { LocalEntryId(UUID.randomUUID().toString()) },
) {
    /** Primary Android import path (ARCHITECTURE §8.4): a SAF folder tree, persisted so future
     *  rescans do not need the picker again. */
    suspend fun importTree(treeUri: Uri) {
        persistPermission(treeUri)
        reconcileAndIndex(SafLibraryScanner.scanTree(context, treeUri))
    }

    /**
     * Explicit multi-select (`ACTION_OPEN_DOCUMENT`, brief §10's "multiple files import"). Unlike
     * [importTree]'s folder walk, a file picked here is indexed **regardless of extension** —
     * `test-media/synthetic/unsupported.xyz` exists exactly to prove this path answers
     * deterministically ([DecodeStatus.UNSUPPORTED]) rather than silently dropping a deliberate
     * user choice the way an incidental non-audio file in a scanned folder is dropped.
     */
    suspend fun importFiles(uris: List<Uri>) {
        val now = monotonicNowUs()
        for (uri in uris) {
            currentCoroutineContext().ensureActive()
            persistPermission(uri)
            val filename = displayNameOf(uri) ?: uri.lastPathSegment ?: uri.toString()
            val sizeBytes = sizeOf(uri) ?: 0L
            indexExplicit(DiscoveredLocation(uri.toString(), filename, sizeBytes), now)
        }
    }

    /** Secondary "whole device library" convenience (ARCHITECTURE §8.4). */
    suspend fun rescanMediaStore() {
        reconcileAndIndex(MediaStoreLibraryScanner.scan(context))
    }

    /**
     * ADR-005's background lazy pass: computes the authoritative
     * [com.ridelink.core.model.ContentHash] for every row that does not have one yet, one file at a
     * time so a cancelled pass leaves a correct, resumable partial result rather than a half-written
     * one. Reads the current set of not-yet-hashed rows directly from the repository — never a
     * possibly-stale caller-supplied snapshot (this phase's closure-audit hardening pass) — so a
     * fresh call always resumes exactly the rows still missing a hash, whether that is because a
     * previous pass was cancelled or because new tracks were imported since.
     */
    suspend fun completeContentHashing() {
        for (entry in repository.entriesMissingContentHash()) {
            currentCoroutineContext().ensureActive()
            val hash =
                runCatching {
                    ContentHashing.computeContentHash(context.contentResolver, Uri.parse(entry.location.uri))
                }.getOrNull() ?: continue
            repository.updateContentHash(entry.localEntryId, hash)
        }
    }

    private suspend fun reconcileAndIndex(discovered: List<DiscoveredLocation>) {
        val now = monotonicNowUs()
        val byLocation = LinkedHashMap<LocalTrackLocation, DiscoveredLocation>()
        val quickIdByLocation = LinkedHashMap<LocalTrackLocation, QuickId>()
        for (location in discovered) {
            currentCoroutineContext().ensureActive()
            val quickId =
                runCatching { ContentHashing.computeQuickId(context.contentResolver, Uri.parse(location.uri)) }.getOrNull()
                    ?: continue // unreadable: skip this scan pass rather than fabricate an identity for it
            val loc = LocalTrackLocation(location.uri)
            byLocation[loc] = location
            quickIdByLocation[loc] = quickId
        }

        val plan = IndexReconciliation.reconcile(repository.allLocationsAndQuickIds(), quickIdByLocation)

        // New and changed locations both go through the same full pipeline; the only difference —
        // whether an existing row's LocalEntryId is preserved or a fresh one is generated — is
        // decided inside indexOrReindex by looking the location up, never by which bucket it came
        // from (this keeps a race between this scan and a concurrent one safe: at worst a location
        // is re-read once more than strictly necessary, never double-inserted, since locationUri is
        // UNIQUE at the schema level).
        for (loc in plan.newLocations + plan.changedLocations) {
            currentCoroutineContext().ensureActive()
            indexOrReindex(byLocation.getValue(loc), quickIdByLocation.getValue(loc), now)
        }
        for (loc in plan.unchangedLocations) {
            currentCoroutineContext().ensureActive()
            repository.touchSeen(loc, now)
        }
        for (loc in plan.missingLocations) {
            currentCoroutineContext().ensureActive()
            repository.markMissing(loc, now)
        }
    }

    private suspend fun indexExplicit(
        location: DiscoveredLocation,
        now: Long,
    ) {
        if (!AudioFormats.isSupportedExtension(location.filename)) {
            upsertPlaceholder(location, now, DecodeStatus.UNSUPPORTED)
            return
        }
        val uri = Uri.parse(location.uri)
        val quickId = runCatching { ContentHashing.computeQuickId(context.contentResolver, uri) }.getOrNull()
        if (quickId == null) {
            upsertPlaceholder(location, now, DecodeStatus.CORRUPT)
            return
        }
        indexOrReindex(location, quickId, now)
    }

    /**
     * Inserts a never-before-seen [location] or, if this exact [LocalTrackLocation] is already
     * known (an in-place edit reconciliation already detected, or a re-import of the same explicit
     * pick), updates that row in place — preserving its [LocalEntryId], never creating a second row
     * for the same location and never deciding identity from [quickId].
     */
    private suspend fun indexOrReindex(
        location: DiscoveredLocation,
        quickId: QuickId,
        now: Long,
    ) {
        val existing = repository.findByLocationUri(location.uri)
        val entry =
            buildEntry(
                existing?.localEntryId ?: newLocalEntryId(),
                location,
                quickId,
                now,
                MetadataExtractor.extract(context, Uri.parse(location.uri)),
            )
        if (existing == null) repository.insertNew(entry) else repository.updateReindexed(entry)
    }

    private suspend fun upsertPlaceholder(
        location: DiscoveredLocation,
        now: Long,
        status: DecodeStatus,
    ) {
        val existing = repository.findByLocationUri(location.uri)
        val entry = placeholderEntry(existing?.localEntryId ?: newLocalEntryId(), location, now, status)
        if (existing == null) repository.insertNew(entry) else repository.updateReindexed(entry)
    }

    private fun buildEntry(
        localEntryId: LocalEntryId,
        location: DiscoveredLocation,
        quickId: QuickId,
        now: Long,
        extraction: Result<ExtractedMetadata>,
    ): LibraryEntry {
        val metadata = extraction.getOrNull()
        // Measured directly on a real emulator (docs/STATUS.md's own discipline of recording what
        // was actually observed): MediaMetadataRetriever does not reliably throw for a severely
        // truncated/malformed container — `setDataSource` can succeed and every `extractMetadata`
        // call simply return null. A zero/absent duration is the signal that actually distinguishes
        // "genuinely unparseable" from "no_metadata.m4a"'s merely-untagged-but-playable case, since
        // a real container MediaMetadataRetriever could open at all reports a real duration even
        // with no title/artist/album tags present.
        val decodeStatus = if (metadata == null || metadata.durationMs <= 0L) DecodeStatus.CORRUPT else DecodeStatus.INDEXED
        val artworkRef =
            metadata
                ?.artworkBytes
                ?.let(ArtworkProcessor::processToBoundedJpeg)
                ?.let { artworkCache.store(localEntryId, it) }
        return LibraryEntry(
            localEntryId = localEntryId,
            track =
                Track(
                    // The authoritative hash is never computed on the fast indexing path (ADR-005) —
                    // completeContentHashing fills this in later, in the background. An in-place edit
                    // resets it to unknown too (the old hash no longer describes the current bytes).
                    contentHash = null,
                    quickId = quickId,
                    title = MetadataNormalizer.title(metadata?.title, location.filename),
                    artist = MetadataNormalizer.artist(metadata?.artist),
                    album = MetadataNormalizer.album(metadata?.album),
                    durationMs = metadata?.durationMs ?: 0L,
                    filename = location.filename,
                    codec = metadata?.codec ?: extensionOf(location.filename),
                    bitrateKbps = metadata?.bitrateKbps ?: 0,
                    artworkRef = artworkRef,
                    sizeBytes = location.sizeBytes,
                ),
            location = LocalTrackLocation(location.uri),
            decodeStatus = decodeStatus,
            indexedAtMonoUs = now,
            lastSeenAtMonoUs = now,
        )
    }

    /**
     * A row for a file that was never hashed at all — [DecodeStatus.UNSUPPORTED] (extension gate,
     * never opened) or [DecodeStatus.CORRUPT] (couldn't even read enough bytes to compute
     * [com.ridelink.core.model.QuickId]). Its `quickId` column is keyed by a hash of the URI itself,
     * **not** a content hash — documented distinctly so nothing downstream mistakes it for real
     * content identity; its real identity is still [localEntryId], exactly like every other row.
     */
    private fun placeholderEntry(
        localEntryId: LocalEntryId,
        location: DiscoveredLocation,
        now: Long,
        status: DecodeStatus,
    ): LibraryEntry =
        LibraryEntry(
            localEntryId = localEntryId,
            track =
                Track(
                    contentHash = null,
                    quickId = syntheticQuickIdForUri(location.uri),
                    title = MetadataNormalizer.title(null, location.filename),
                    artist = MetadataNormalizer.artist(null),
                    album = MetadataNormalizer.album(null),
                    durationMs = 0,
                    filename = location.filename,
                    codec = extensionOf(location.filename),
                    bitrateKbps = 0,
                    artworkRef = null,
                    sizeBytes = location.sizeBytes,
                ),
            location = LocalTrackLocation(location.uri),
            decodeStatus = status,
            indexedAtMonoUs = now,
            lastSeenAtMonoUs = now,
        )

    private fun syntheticQuickIdForUri(uri: String): QuickId {
        val digest = MessageDigest.getInstance("SHA-256").digest(uri.toByteArray(Charsets.UTF_8))
        return QuickId("sha256:" + digest.joinToString("") { "%02x".format(it) })
    }

    private fun extensionOf(filename: String): String = filename.substringAfterLast('.', missingDelimiterValue = "")

    private fun persistPermission(uri: Uri) {
        // Best-effort: a provider that does not support persistable permissions (some MediaStore
        // rows) throws here, and the file is still usable for this session either way.
        runCatching {
            context.contentResolver.takePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }

    private fun displayNameOf(uri: Uri): String? =
        context.contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
            if (cursor.moveToFirst()) cursor.getString(0) else null
        }

    private fun sizeOf(uri: Uri): Long? =
        context.contentResolver.query(uri, arrayOf(OpenableColumns.SIZE), null, null, null)?.use { cursor ->
            if (cursor.moveToFirst() && !cursor.isNull(0)) cursor.getLong(0) else null
        }
}

File: ios/Packages/RideLinkPlatform/Sources/RideLinkPlatform/Library/LibraryIndexer.swift
import CryptoKit
import Foundation
import RideLinkCore

/// A file the document picker returned, before hashing or metadata extraction — deliberately not a
/// domain type (`LibraryEntry` requires a `Track`, which needs a hash this stage does not have yet).
/// Mirrors `com.ridelink.data.library.DiscoveredLocation`.
struct DiscoveredLocation {
    let sourceUrl: URL
    let filename: String
}

/// The whole indexing pipeline (this phase's brief §19): discover → validate supported type →
/// look up any existing row for this location → copy into the app's own container → `QuickId` →
/// metadata → artwork → insert/update. Every step that can fail does so into a `DecodeStatus`, never
/// a thrown error the caller has to catch — "even local files are untrusted input" (brief §20)
/// applies to every file this touches.
///
/// **Identity (ADR-005 Amendment A1, this phase's closure-audit CRITICAL finding):** every row's real
/// identity is a freshly-generated `LocalEntryId` — `newLocalEntryId` by default, overridable only for
/// deterministic tests — preserved across a reimport of the same location, never recomputed from
/// content. `location.uri` (`LibraryEntry.location`, stored in the `locationUri` column) is the
/// **source URL** the picker/enumerator returned, not the app-container destination path: it is what
/// `LibraryRepository.findByLocationUri` looks up to decide "have we already imported this exact
/// pick?", exactly the role Android's persisted `content://` URI plays as the schema's `UNIQUE`
/// location key. The app-container destination filename is derived from `LocalEntryId` instead
/// (`resolvedUrl(for:)`) — **never from `QuickId`**, which is not guaranteed unique across rows: two
/// genuinely different files over 128 KiB can share one, and deriving a destination filename from it
/// used to mean a colliding file's *bytes were never even copied* into the sandbox, because import
/// skipped copying to a destination that already existed. A reimport of the same source location now
/// always removes and recopies its destination file in place, never skip-if-exists.
///
/// **A real, deliberate divergence from `com.ridelink.data.library.LibraryIndexer`, not an
/// oversight**: Android's SAF import keeps a persisted `content://` permission and re-scans the
/// *original* location every time, so a file removed from that location is detected as
/// `DecodeStatus.missing` (`IndexReconciliation`). iOS never does that — ADR-009 says "no reliance
/// on an external security-scoped URL indefinitely", so every picked file is copied into this app's
/// own `Application Support` directory on import and the security-scoped source URL is never touched
/// again. There is therefore no live external reference whose disappearance this indexer could
/// detect; `IndexReconciliation`'s `missingLocations` path (used by Android's `SafLibraryScanner`) has
/// no iOS caller for that reason, not because it was forgotten. Re-importing the same *source
/// location* is a no-op update that preserves identity — the same de-duplication-of-the-same-pick
/// FR-010 requires on both platforms — but Phase 3 does **not** collapse two different source
/// locations that happen to share a `QuickId` or even a `ContentHash`; that is Phase 4/5 transfer
/// scope (ADR-005 Amendment A1).
public final class LibraryIndexer: Sendable {
    private let repository: LibraryRepository
    private let artworkCache: ArtworkCache
    private let musicDirectory: URL
    private let monotonicNowUs: @Sendable () -> Int64
    private let newLocalEntryId: @Sendable () -> LocalEntryId

    public init(
        repository: LibraryRepository,
        artworkCache: ArtworkCache,
        musicDirectory: URL,
        monotonicNowUs: @escaping @Sendable () -> Int64,
        newLocalEntryId: @escaping @Sendable () -> LocalEntryId = { LocalEntryId(UUID().uuidString.lowercased()) }
    ) {
        self.repository = repository
        self.artworkCache = artworkCache
        self.musicDirectory = musicDirectory
        self.monotonicNowUs = monotonicNowUs
        self.newLocalEntryId = newLocalEntryId
    }

    /// Explicit multi-select (`UIDocumentPickerViewController`, brief §10's "multiple files import").
    /// A file picked here is indexed **regardless of extension** — mirrors Android's `importFiles`:
    /// `DecodeStatus.unsupported` answers deterministically rather than silently dropping a
    /// deliberate user choice.
    public func importFiles(_ urls: [URL]) async throws {
        try ensureMusicDirectoryExists()
        for url in urls {
            try Task.checkCancellation()
            try await importOne(DiscoveredLocation(sourceUrl: url, filename: url.lastPathComponent), enforceExtensionGate: false)
        }
    }

    /// Folder import (ARCHITECTURE §8.4's iOS path): recursively enumerates `folderUrl` and imports
    /// every file whose extension `AudioFormats` recognizes, silently skipping the rest — the folder
    /// itself was not a deliberate "index this exact file" choice the way a multi-select pick is, so
    /// an incidental non-audio file inside it is dropped rather than recorded as `.unsupported`
    /// (matches Android's `SafLibraryScanner`/`importTree` behaviour for a scanned tree).
    public func importFolder(_ folderUrl: URL) async throws {
        try ensureMusicDirectoryExists()
        let accessing = folderUrl.startAccessingSecurityScopedResource()
        defer { if accessing { folderUrl.stopAccessingSecurityScopedResource() } }

        let enumerator = FileManager.default.enumerator(
            at: folderUrl, includingPropertiesForKeys: [.isRegularFileKey], options: [.skipsHiddenFiles]
        )
        var files: [URL] = []
        while let candidate = enumerator?.nextObject() as? URL {
            try Task.checkCancellation()
            let isRegularFile = try candidate.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile ?? false
            if isRegularFile, AudioFormats.isSupportedExtension(candidate.lastPathComponent) {
                files.append(candidate)
            }
        }
        for url in files {
            try Task.checkCancellation()
            try await importOne(DiscoveredLocation(sourceUrl: url, filename: url.lastPathComponent), enforceExtensionGate: true)
        }
    }

    /// ADR-005's background lazy pass: computes the authoritative `ContentHash` for every row that
    /// does not have one yet, one file at a time so a cancelled pass leaves a correct, resumable
    /// partial result rather than a half-written one. **No-arg** (this phase's closure-audit
    /// hardening pass, Finding B): reads the current set of not-yet-hashed rows directly from the
    /// repository — never a possibly-stale caller-supplied snapshot — so a fresh call always resumes
    /// exactly the rows still missing a hash, whether that is because a previous pass was cancelled or
    /// because new tracks were imported since.
    public func completeContentHashing() async throws {
        for entry in try repository.entriesMissingContentHash() {
            try Task.checkCancellation()
            guard let hash = try? ContentHashing.computeContentHash(fileURL: resolvedUrl(for: entry)) else { continue }
            try? repository.updateContentHash(localEntryId: entry.localEntryId, contentHash: hash)
        }
    }

    /// Resolves this row's app-container destination file — derived from its `LocalEntryId`, never
    /// from `entry.location.uri` (which stores the original *source* URL, not a durably-reopenable
    /// reference; see this type's own doc comment). The extension is taken from `Track.filename`
    /// (the original display name), matching whatever `importOne` used when it copied the bytes in.
    public func resolvedUrl(for entry: LibraryEntry) -> URL {
        musicDirectory.appendingPathComponent(destinationFilename(localEntryId: entry.localEntryId, filename: entry.track.filename))
    }

    private func importOne(_ location: DiscoveredLocation, enforceExtensionGate: Bool) async throws {
        let now = monotonicNowUs()
        let locationUri = location.sourceUrl.absoluteString
        let existing = try? repository.findByLocationUri(locationUri)

        if enforceExtensionGate == false, !AudioFormats.isSupportedExtension(location.filename) {
            try? save(placeholderEntry(location, locationUri: locationUri, existing: existing, now: now, status: .unsupported), isNew: existing == nil)
            return
        }

        let accessing = location.sourceUrl.startAccessingSecurityScopedResource()
        defer { if accessing { location.sourceUrl.stopAccessingSecurityScopedResource() } }

        guard let quickId = try? ContentHashing.computeQuickId(fileURL: location.sourceUrl) else {
            try? save(placeholderEntry(location, locationUri: locationUri, existing: existing, now: now, status: .corrupt), isNew: existing == nil)
            return
        }

        let localEntryId = existing?.localEntryId ?? newLocalEntryId()
        let destinationUrl = musicDirectory.appendingPathComponent(destinationFilename(localEntryId: localEntryId, filename: location.filename))
        // Never skip-if-exists (ADR-005 Amendment A1's iOS CRITICAL finding): a destination file can
        // only already exist here when `existing` preserved this exact location's `LocalEntryId`, in
        // which case a reimport must overwrite it with fresh bytes — leaving stale bytes in place
        // because a file of that name happened to exist was the whole bug.
        try? FileManager.default.removeItem(at: destinationUrl)
        guard (try? FileManager.default.copyItem(at: location.sourceUrl, to: destinationUrl)) != nil else {
            try? save(placeholderEntry(location, locationUri: locationUri, existing: existing, now: now, status: .corrupt), isNew: existing == nil)
            return
        }

        let sizeBytes = (try? destinationUrl.resourceValues(forKeys: [.fileSizeKey]).fileSize.map(Int64.init)) ?? nil
        let extraction = await MetadataExtractor.extract(fileURL: destinationUrl)
        let entry = buildEntry(
            localEntryId: localEntryId, location: location, quickId: quickId, locationUri: locationUri,
            sizeBytes: sizeBytes ?? 0, now: now, extraction: extraction
        )
        try? save(entry, isNew: existing == nil)
    }

    /// Inserts a never-before-seen location, or, if this exact source location is already known,
    /// updates that row in place — preserving its `LocalEntryId`, never creating a second row for the
    /// same location.
    private func save(_ entry: LibraryEntry, isNew: Bool) throws {
        if isNew {
            try repository.insertNew(entry)
        } else {
            try repository.updateReindexed(entry)
        }
    }

    private func buildEntry(
        localEntryId: LocalEntryId,
        location: DiscoveredLocation,
        quickId: QuickId,
        locationUri: String,
        sizeBytes: Int64,
        now: Int64,
        extraction: Result<ExtractedMetadata, Error>
    ) -> LibraryEntry {
        let metadata = try? extraction.get()
        // The same measured heuristic Android's indexer uses: some malformed containers are opened
        // without error yet report no usable duration, so `durationMs <= 0` is the second CORRUPT
        // signal alongside a thrown extraction error.
        let decodeStatus: DecodeStatus = (metadata == nil || (metadata?.durationMs ?? 0) <= 0) ? .corrupt : .indexed
        let artworkRef = metadata?.artworkData.flatMap(ArtworkProcessor.processToBoundedJpeg).flatMap {
            artworkCache.store(localEntryId: localEntryId, data: $0)
        }
        return LibraryEntry(
            localEntryId: localEntryId,
            track: Track(
                // The authoritative hash is never computed on the fast indexing path (ADR-005) —
                // completeContentHashing fills this in later, in the background. A reimport of an
                // in-place edit resets it to unknown too (the old hash no longer describes the
                // current bytes) via `LibraryRepository.updateReindexed`.
                contentHash: nil,
                quickId: quickId,
                title: MetadataNormalizer.title(metadata?.title, filename: location.filename),
                artist: MetadataNormalizer.artist(metadata?.artist),
                album: MetadataNormalizer.album(metadata?.album),
                durationMs: metadata?.durationMs ?? 0,
                filename: location.filename,
                codec: metadata?.codec ?? extensionOf(location.filename),
                bitrateKbps: metadata?.bitrateKbps ?? 0,
                artworkRef: artworkRef,
                sizeBytes: sizeBytes
            ),
            location: LocalTrackLocation(uri: locationUri),
            decodeStatus: decodeStatus,
            indexedAtMonoUs: now,
            lastSeenAtMonoUs: now
        )
    }

    /// A row for a file that was never hashed at all (`DecodeStatus.unsupported`, extension gate,
    /// never opened; or `.corrupt`, couldn't even read enough bytes to compute a `QuickId`, or the
    /// app-owned copy failed). Its `quickId` is a hash of the source URL string itself, **not** a
    /// content hash — documented distinctly so nothing downstream mistakes it for real content
    /// identity; its real identity is still `localEntryId`, exactly like every other row.
    private func placeholderEntry(
        _ location: DiscoveredLocation,
        locationUri: String,
        existing: LibraryEntry?,
        now: Int64,
        status: DecodeStatus
    ) -> LibraryEntry {
        LibraryEntry(
            localEntryId: existing?.localEntryId ?? newLocalEntryId(),
            track: Track(
                contentHash: nil,
                quickId: syntheticQuickId(for: locationUri),
                title: MetadataNormalizer.title(nil, filename: location.filename),
                artist: MetadataNormalizer.artist(nil),
                album: MetadataNormalizer.album(nil),
                durationMs: 0,
                filename: location.filename,
                codec: extensionOf(location.filename),
                bitrateKbps: 0,
                artworkRef: nil,
                sizeBytes: 0
            ),
            location: LocalTrackLocation(uri: locationUri),
            decodeStatus: status,
            indexedAtMonoUs: now,
            lastSeenAtMonoUs: now
        )
    }

    private func destinationFilename(localEntryId: LocalEntryId, filename: String) -> String {
        let ext = (filename as NSString).pathExtension
        return ext.isEmpty ? localEntryId.value : "\(localEntryId.value).\(ext)"
    }

    private func syntheticQuickId(for uriString: String) -> QuickId {
        let digest = SHA256.hash(data: Data(uriString.utf8))
        return QuickId("sha256:" + digest.map { String(format: "%02x", $0) }.joined())
    }

    private func extensionOf(_ filename: String) -> String { (filename as NSString).pathExtension }

    private func ensureMusicDirectoryExists() throws {
        try FileManager.default.createDirectory(at: musicDirectory, withIntermediateDirectories: true)
    }
}
